Bug 1640201 - Add the high level design docs for the upload module [doc only] (#1056)
mozilla/glean: docs/dev/core/internal/upload.md

Glean core controls all upload and coordinates the platform side with its own internals.
All language bindings implement ping uploading around a common API and protocol.

## The upload module in the language bindings

> This paragraph illustrates the recommended structure of the upload module implementation in the language bindings (regardless of the target language), in order to allow greater flexibility with respect to potential future requirements and features.

Reviewer: I don't think this "intent sentence" is necessary.

```mermaid
classDiagram
    class UploadResult {
        ~ToFFI()* int
    }

    class HttpResponse {
        int statusCode
        ~ToFFI() int
    }

    class UnrecoverableFailure {
        ~ToFFI() int
    }

    class RecoverableFailure {
        ~ToFFI() int
    }

    class PingUploader {
        <<interface>>
        +Upload() UploadResult
    }

    class BaseUploader {
        +BaseUploader(PingUploader)
        ~TriggerUploads()
        ~CancelUploads()
    }

    class HttpUploader {
        +Upload() UploadResult
    }

    UploadResult <|-- HttpResponse
    UploadResult <|-- UnrecoverableFailure
    UploadResult <|-- RecoverableFailure
    PingUploader <|-- HttpUploader
    PingUploader o-- BaseUploader

> ```

Reviewer: Is there a way to provide a text caption to this figure?

Author: Not that I know of, I'm afraid :(

Reviewer 2: Since markdown accepts any valid HTML, you could put this inside a [`figure`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figure) tag and provide a `figcaption` right next to it. Apparently mermaid creates an `img` tag where all this code is, so that would work perfectly. I could be wrong as to how mermaid works though.

Author: This doesn't seem to work, at least locally, for me :\ The mermaid doesn't seem to get parsed?

Reviewer 2: Hum, oh well. Maybe we can open a follow up to look into this with more time. I see the other upload graph we have also doesn't have a caption so it could be a more broad thing where we add captios to all images in the book.

Author: Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1652705

* The `PingUploader` interface describes the contract between the `BaseUploader` and the SDK or user-provided upload modules.

> * There is a `BaseUploader` component that is responsible for interfacing with the lower level FFI `get_upload_*` calls and dealing with the logic in a platform-coherent way.

Reviewer 2: ```suggestion
* There is a `BaseUploader` component that is responsible for interfacing with the lower level FFI `get_upload_task` calls and dealing with the logic in a platform-coherent way.
```

There is just the `get_upload_task` function and no other "getter"s. No need for asterisk I think.

Reviewer: Also, not all language bindings use FFI calls, so I think we can skip the word `FFI` here.

  * Every Glean instance will always have a single `BaseUploader` instance.
  * The `BaseUploader` is fed, at Glean initialization, with an instance of an implementation of the `PingUploader` interface.
  * Whenever `BaseUploader` thinks it should perform an upload, it will call the provided instance of the `PingUploader` interface and call `upload` with the data it's getting from the glean-core/FFI.

>   * Any throttling happens at this layer!

Reviewer 2: ```suggestion
  * Any throttling happens at this layer.
```
Any specific reason for the exclamation here? One more thing, do you think it is necessary to point out that the core is what should orchestrate the throttling and the binding will just abide to what the core is telling it to do?

Author: Oh, I tend to get overexcited with punctuation! Sorry about that!

Yes, it makes sense to clarify the throttling part. Let me add that.

  * Any logic for aborting uploads or triggering uploads is provided by this object.
* The `HttpClientUploader` is the default SDK-provided HTTP uploader. It acts as an adapter between the platform-specific upload library and the Glean upload APIs.

> > Note that most of the language have now diverged, due to the many iterations that got us here, from this design. For example, in Kotlin, the [`BaseUploader` is mostly empty](https://searchfox.org/glean/source/glean-core/android/src/main/java/mozilla/telemetry/glean/net/BaseUploader.kt) and its functionalities are spread in the [`PingUploadWorker`](https://searchfox.org/glean/source/glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt).

Reviewer 2: ```suggestion
> Note that most of the languages have now diverged, due to the many iterations, from this design. For example, in Kotlin, the [`BaseUploader` is mostly empty](https://searchfox.org/glean/source/glean-core/android/src/main/java/mozilla/telemetry/glean/net/BaseUploader.kt) and its functionalities are spread in the [`PingUploadWorker`](https://searchfox.org/glean/source/glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt).
```

Glean core controls all upload and coordinates the platform side with its own internals.
All language bindings implement ping uploading around a common API and protocol.

## The upload module in the language bindings

```mermaid
classDiagram
    class UploadResult {
        ~ToFFI()* int
    }

    class HttpResponse {
        int statusCode
        ~ToFFI() int
    }

    class UnrecoverableFailure {
        ~ToFFI() int
    }

    class RecoverableFailure {
        ~ToFFI() int
    }

    class PingUploader {
        <<interface>>
        +Upload() UploadResult
    }

    class BaseUploader {
        +BaseUploader(PingUploader)
        ~TriggerUploads()
        ~CancelUploads()
    }

    class HttpUploader {
        +Upload() UploadResult
    }

    UploadResult <|-- HttpResponse
    UploadResult <|-- UnrecoverableFailure
    UploadResult <|-- RecoverableFailure
    PingUploader <|-- HttpUploader
    PingUploader o-- BaseUploader

```

### `PingUploader`
The `PingUploader` interface describes the contract between the `BaseUploader` and the SDK or user-provided upload modules.

### `BaseUploader`

> The `BaseUploader` component that is responsible for interfacing with the lower level `get_upload_task` calls and dealing with the logic in a platform-coherent way.

Reviewer: ```suggestion
The `BaseUploader` component is responsible for interfacing with the lower level `get_upload_task` calls and dealing with the logic in a platform-coherent way.
```
